### PR TITLE
Brand new geometry editor' erase tool functionality

### PR DESCRIPTION
**Pull request by nirvn**
Exciting screencast time:

[Screencast from 2023-09-18 15-57-01.webm](https://github.com/opengisch/QField/assets/1728657/586bb38d-1608-4c54-a91b-9ad65d6cecef)

This PR adds a brand new geometry editor functionality: an erase tool. The tool allows users to "erase" part of a {line, polygon} geometry. 

~~(The PR temporarily includes the commits from https://github.com/opengisch/QField/pull/4579)~~

### CODE REVIEW

**Comment by nirvn** [comment]
(This has been rebased and is ready for review)

**Review by suricactus — COMMENTED** [review]
Some small comments on the code.

File: src/core/utils/geometryutils.cpp [hunk 1]
```diff
@@ -98,6 +106,30 @@ GeometryUtils::GeometryOperationResult GeometryUtils::reshapeFromRubberband( Qgs
   return reshapeReturn;
 }
 
+GeometryUtils::GeometryOperationResult GeometryUtils::eraseFromRubberband( QgsVectorLayer *layer, QgsFeatureId fid, RubberbandModel *rubberBandModel )
+{
+  QgsFeature feature = layer->getFeature( fid );
+  QgsGeometry geom = feature.geometry();
+  if ( geom.isNull() || ( QgsWkbTypes::geometryType( geom.wkbType() ) != Qgis::GeometryType::Line && QgsWkbTypes::geometryType( geom.wkbType() ) != Qgis::GeometryType::Polygon ) )
+    return GeometryUtils::GeometryOperationResult::InvalidBaseGeometry;
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/core/utils/geometryutils.cpp:114]: @suricactus . have we settled on this empty `if (` line styling or not? I'm not liking that styling (and it's also not consistent with QGIS source code). I'm +1 on avoiding long conditional lines, but -1 on condition-free if ( lines.

File: src/qml/geometryeditors/Erase.qml [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+      rubberbandModel.addVertex()
+    }
+
+    onConfirmed: {
+      digitizingLogger.writeCoordinates()
+
+      rubberbandModel.frozen = true
+      if (!featureModel.currentLayer.editBuffer())
+          featureModel.currentLayer.startEditing()
```
[the remark this answers is not in the document]
Comment (nirvn) [reply src/qml/geometryeditors/Erase.qml:71]: @suricactus , thanks, I need to drill myself into avoiding these {}-less blocks

**Comment by nirvn** [comment]
@suricactus , comments addressed (inc. those in the chat about multipart issue).

**Review by suricactus — APPROVED** [review]
Code looks good. Have not tried myself yet, will be able later this week if needed.

**Pull request merged**